Source/WebCore/svg/SVGCircleElement.cpp

```diff
@@ -51,6 +51,11 @@ Ref<SVGCircleElement> SVGCircleElement::create(const QualifiedName& tagName, Doc
     return adoptRef(*new SVGCircleElement(tagName, document));
 }
 
+bool SVGCircleElement::selfHasRelativeLengths() const
+{
+    return cx().isRelative() || cy().isRelative() || r().isRelative();
+}
```

Comment:
This is approach is okay. But I think it can be done through the `SVGPropertyRegistry`.

You can add a virtual function named `selfHasRelativeLengths()` in `SVGPropertyRegistry`. `SVGPropertyOwnerRegistry` can override this function like this:
```
bool selfHasRelativeLengths() const override
{
    for (const auto& entry : attributeNameToAccessorMap()) {
        if (entry.value->isRelativeLength(m_owner))
            return true;
    }
    return false;
}
```

Where `isRelativeLength()` is a virtual function defined in `SVGMemberAccessor` and it is overridden by `SVGAnimatedLengthAccessor` like this:
```
bool isRelativeLength(const OwnerType& owner) const override { return property(owner)->currentValue().isRelative(); }
```

So `SVGElement::selfHasRelativeLengths()` does not have to be virtual anymore and can be changed to be like this:
```
bool SVGElement::selfHasRelativeLengths() const
{
    return propertyRegistry().selfHasRelativeLengths();
}
```

Reply:
I have a test failure that seems definitely related to my change so I must be doing something wrong but no idea what.

Reply:
> So SVGElement::selfHasRelativeLengths() does not have to be virtual anymore and can be changed to be like this:

I don't think it's that easy, some `selfHasRelativeLengths()` overrides are not that trivial.

Reply:
I think you should keep the `SVGRectElement::selfHasRelativeLengths()` return true as is.

Reply:
Can you explain why though? I have no doubt it will fix the test but I am not understanding why my new implementation is not sufficient.

Reply:
I do not think there is a good explanation other than it was easier to do it that way. If we want to know to for sure whether the SVGRectElement has a relative length or not, it will be more difficult.

The test has this SVG
```
<svg width="400" height="200">
  <rect width="200" height="200" fill="red"/>
  <rect id="target" x="100" width="200" height="200" fill="green"/>
</svg>
```
And the second rect (id="target") has this style

```
<style>
#target {
  transform-box: view-box;
  transform: translate(-50%, 0);
}
</style>
```
In JS we change the width of the SVG 
```
    document.querySelector('svg').setAttribute('width', 200);
```

So the none of SVGRectElement attributes is a relative SVGAnimatedLength. But the css transform itself as a relative translate.

Reply:
Wouldn't other types than SVGRectElement have the same issue when applied such css though?

Reply:
I think all the ones that have location and size return true right now.

Reply:
I think the reason for this confusion is the name selfHasRelativeLengths(). Maybe we need to change it to `needsUpdateRelativeLengthsInformation()` or something similar.

Reply:
Blink is able to do what I did in my patch though so I suspect we have a bug somewhere else.